[PEP518] Improve testsuite packaging

IAlibay opened the pull request:
Changes made in this Pull Request:
 - Adds pyproject.toml to testsuite
 - Clears up ambigous packaging
 - Keeps setup.py for dynamic things only
 - Update license file
 - Update links info
 - Update some setup.py docstring text
 - Adds various missing init files


PR Checklist
------------
 - Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - Issue raised/referenced?

## Developers certificate of origin
- [x] I certify that this contribution is covered by the LGPLv2.1+ license as defined in our [LICENSE](https://github.com/MDAnalysis/mdanalysis/blob/develop/LICENSE) and adheres to the [**Developer Certificate of Origin**](https://developercertificate.org/).


<!-- readthedocs-preview mdanalysis start -->
----
📚 Documentation preview 📚: https://mdanalysis--4463.org.readthedocs.build/en/4463/

<!-- readthedocs-preview mdanalysis end -->

IAlibay committed:
Update testsuite packaging

IAlibay committed:
Add pytest ini options to pyproject.toml

pep8speaks commented:
Hello @IAlibay! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`testsuite/setup.py`](https://github.com/MDAnalysis/mdanalysis/blob/7b69052ddf7344dc9c1757d39cfeda1ba082cda7/testsuite/setup.py):

> [Line 83:29](https://github.com/MDAnalysis/mdanalysis/blob/7b69052ddf7344dc9c1757d39cfeda1ba082cda7/testsuite/setup.py#L83): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment



##### Comment last updated at 2024-02-23 07:10:18 UTC

IAlibay committed:
update changelog

IAlibay committed:
Update pyproject.toml

IAlibay committed:
Add some extra init files

RMeli commented on package/pyproject.toml:
```diff
@@ -97,11 +97,13 @@ doc = [
-"User Group" = 'https://groups.google.com/g/mdnalysis-discussion/'
-Discord = 'https://discord.com/channels/807348386012987462/'
-Blog = 'https://www.mdanalysis.org/blog/'
+"Project URL" = "https://www.mdanalysis.org"
+"Download URL" = "https://github.com/MDAnalysis/mdanalysis/releases"
```
See https://github.com/MDAnalysis/mdanalysis/pull/4424#discussion_r1499912898

IAlibay replied:
Updated

RMeli commented on package/pyproject.toml:
```diff
@@ -97,11 +97,13 @@ doc = [
+"Github Discussions" = "https://github.com/MDAnalysis/mdanalysis/discussions"
+Discord = "https://discord.com/channels/807348386012987462/"
+Blog = "https://www.mdanalysis.org/blog/"
 Twitter = 'https://twitter.com/mdanalysis'
```
Add LinkedIn? And other platforms?

Also, Twitter is no more... Should this be changed to X?

IAlibay replied:
Would it be better if we just dropped socials completely? It feels a bit silly to have to keep this up to date for just the pypi metadata

IAlibay replied:
I removed twitter both here and the other PR, I'm hoping the rest should be stable enough.

RMeli replied:
Yes, makes sense to remove socials!

RMeli commented on testsuite/MDAnalysisTests/parallelism/__init__.py:
Nitpick: the other `__init__.py` files all contain one (empty) line, while this is empty. The ones in https://github.com/MDAnalysis/mdanalysis/pull/4424 are all empty.

Just pointing this out because if we will enable `pre-commit` hooks one day, the empty files will be changed, I think.

IAlibay replied:
I think according to pep8 they should be empty files. I'll go ahead and update everything to match.

IAlibay replied:
done

RMeli commented on testsuite/pyproject.toml:
```diff
@@ -0,0 +1,170 @@
+[build-system]
+requires = [
+  "setuptools >= 40.9.0",
```
Is it correct that MDAnalysis itself (https://github.com/MDAnalysis/mdanalysis/pull/4424) doesn't require a minimum version?

IAlibay replied:
It probably does (because older versions don't support a lot of toml stuff), but we never set it. I might just set it to this value for package too - it's a good "you'll probably have trouble if you go below this" value.

IAlibay replied:
Updated in other PR - now we have a bare minimum pin of 40.9 which is the release that started supporting pyproject.toml.

RMeli commented on testsuite/setup.py:
```diff
@@ -77,122 +72,16 @@ def run(self):
-          )
+    setup(
+        version=RELEASE,
+        install_requires=[
+            'MDAnalysis=={0!s}'.format(RELEASE),  # same as this release!
+            'pytest>=3.3.0', # Raised to 3.3.0 due to Issue 2329
+            'hypothesis',
+        ],
```
Does this remains here because of the requirement that the version match MDAnalysis?

IAlibay replied:
Yes, anything with dynamic matching must be in setup.py

IAlibay committed:
update links

IAlibay committed:
update all init files to be empty

IAlibay commented:
Thanks for the review @RMeli - I think I've addressed most comments, the social media stuff might need more discussion. I'm of the opinion that the pyproject.toml url links barely get used so we could probably just not think about them too much.

RMeli reviewed (approved):
Thanks Irfan, LGTM

IAlibay committed:
Merge branch 'develop' into testsuite-pep518
